sound/soc/intel/boards/sof_board_helpers.c

```diff
@@ -248,8 +252,138 @@ int sof_intel_board_set_intel_hdmi_link(struct device *dev,
 }
 EXPORT_SYMBOL_NS(sof_intel_board_set_intel_hdmi_link, SND_SOC_INTEL_SOF_BOARD_HELPERS);
 
+int sof_intel_board_set_amp_link(struct device *dev,
+				 struct snd_soc_dai_link *link, int be_id,
+				 enum sof_ssp_codec amp_type, int ssp_amp)
+{
+	struct snd_soc_dai_link_component *cpus;
+
+	dev_dbg(dev, "link %d: amp %s, ssp %d\n", be_id,
+		sof_ssp_get_codec_name(amp_type), ssp_amp);
+
+	/* link name */
+	link->name = devm_kasprintf(dev, GFP_KERNEL, "SSP%d-Codec", ssp_amp);
+	if (!link->name)
+		return -ENOMEM;
+
+	/* cpus */
+	cpus = devm_kzalloc(dev, sizeof(struct snd_soc_dai_link_component),
+			    GFP_KERNEL);
+	if (!cpus)
+		return -ENOMEM;
+
+	cpus->dai_name = devm_kasprintf(dev, GFP_KERNEL, "SSP%d Pin", ssp_amp);
+	if (!cpus->dai_name)
+		return -ENOMEM;
+
+	link->cpus = cpus;
+	link->num_cpus = 1;
+
+	/* codecs */
+	switch (amp_type) {
+	case CODEC_CS35L41:
+		cs35l41_set_dai_link(link);
+		break;
+	case CODEC_MAX98357A:
+		max_98357a_dai_link(link);
+		break;
+	case CODEC_MAX98360A:
+		max_98360a_dai_link(link);
+		break;
+	case CODEC_MAX98373:
+		link->codecs = max_98373_components;
+		link->num_codecs = ARRAY_SIZE(max_98373_components);
+		link->init = max_98373_spk_codec_init;
+		link->ops = &max_98373_ops;
+		break;
+	case CODEC_MAX98390:
+		max_98390_dai_link(dev, link);
+		break;
+	case CODEC_NAU8318:
+		nau8318_set_dai_link(link);
+		break;
+	case CODEC_RT1011:
+		sof_rt1011_dai_link(link);
+		break;
+	case CODEC_RT1015:
+		sof_rt1015_dai_link(link);
+		break;
+	case CODEC_RT1015P:
+		sof_rt1015p_dai_link(link);
+		break;
+	case CODEC_RT1019P:
+		sof_rt1019p_dai_link(link);
+		break;
+	case CODEC_RT1308:
+		sof_rt1308_dai_link(link);
+		break;
+	case CODEC_RT5650:
+		/* special case for all-in-one codec, caller to handle */
+		break;
```

Comment:
I think this change probably exposes randconfig problems. If you add these codec-specific parts in SND_SOC_INTEL_SOF_BOARD_HELPERS then you need to do a select ... of all the required codecs. Currently this is only added on a board-by-board basis.

IOW, you need this:

````
config SND_SOC_INTEL_SOF_BOARD_HELPERS
	tristate
        select SND_SOC_INTEL_SOF_MAXIM_COMMON
	select SND_SOC_INTEL_SOF_NUVOTON_COMMON
	select SND_SOC_INTEL_SOF_REALTEK_COMMON
        select SND_SOC_INTEL_SOF_CIRRUS_COMMON
````

And arguably this is the main issue with this PR, we would now have to select all codecs even when they are not used by a distro.

Can we instead pass the sof_codec_dai_link() function as an argument to avoid this switch case and avoid bundling everything together? Same for set_codec_conf(), centralizing everything is not very good in terms of modularity. We should let users or distros the ability to select which board they want, and not bundle everything together IMHO.

Reply:
What I am trying to do is moving all amp support code to the intal-board module so we could add amp support once and for all machine drivers. Thanks for pointing out the module loading problem and I'll do experiment to observe the behavior of run-time kernel module loading and see if we could avoid it.

Reply:
A new version was uploaded to left codec-specific part in machine drivers.